Gem/Code/Source/Components/Multiplayer/GemSpawnerComponent.cpp

```diff
@@ -0,0 +1,263 @@
+            LmbrCentral::Tags tags;
+            LmbrCentral::TagComponentRequestBus::EventResult(tags, gemSpawnEntity,
+                &LmbrCentral::TagComponentRequestBus::Events::GetTags);
+
+            // Randomly select a gem type for this entity.
+            const AZ::Crc32 type = ChooseGemType(tags);
+
+            AZ::Vector3 position = AZ::Vector3::CreateZero();
+            AZ::TransformBus::EventResult(position, gemSpawnEntity, &AZ::TransformBus::Events::GetWorldTranslation);
+            SpawnGem(position, type);
+        }
+    }
+
+    void GemSpawnerComponentController::SpawnGem(const AZ::Vector3& location, const AZ::Crc32& type)
+    {
+        if (GetParent().GetGemSpawnables().empty())
+        {
+            return;
+        }
+
+        const GemSpawnable* spawnable = nullptr;
+        for (const GemSpawnable& gemType : GetParent().GetGemSpawnables())
+        {
+            if (type == AZ::Crc32(gemType.m_tag.c_str()))
+            {
+                spawnable = &gemType;
+            }
+        }
+        if (!spawnable)
+        {
+            return;
+        }
+
+        PrefabCallbacks callbacks;
+        callbacks.m_onActivateCallback = [this](AZStd::shared_ptr<AzFramework::EntitySpawnTicket> ticket,
+            AzFramework::SpawnableConstEntityContainerView view)
+        {
+            for (const AZ::Entity* entity : view)
+            {
+                if (GemComponent* gem = entity->FindComponent<GemComponent>())
+                {
+                    static_cast<GemComponentController*>(gem->GetController())->SetRandomPeriodOffset(
+                        GetNetworkRandomComponentController()->GetRandomInt() % 1000);
+
+                    // Save the gem spawn ticket, otherwise the gem will de-spawn
+                    m_spawnedGems.emplace(entity->GetId(), AZStd::move(ticket));
+                    break;
+                }
+            }
+        };
+
+        GetParent().GetNetworkPrefabSpawnerComponent()->SpawnPrefabAsset(
+            AZ::Transform::CreateFromQuaternionAndTranslation(AZ::Quaternion::CreateIdentity(), location),
+            spawnable->m_gemAsset, AZStd::move(callbacks));
+    }
+
+    void GemSpawnerComponentController::RemoveGems()
+    {
+        for (const auto& gem : m_spawnedGems)
+        {
+            const Multiplayer::NetEntityId netEntityId = Multiplayer::GetNetworkEntityManager()->GetNetEntityIdById(gem.first);
+            Multiplayer::ConstNetworkEntityHandle netEntityHandle = Multiplayer::GetNetworkEntityManager()->GetEntity(netEntityId);
+
+            if (netEntityHandle.Exists())
+            {
+                // Currently blocked by https://github.com/o3de/o3de/issues/11695
+                /*AZ::EntityBus::MultiHandler::BusConnect(gem.first);
+                Multiplayer::GetNetworkEntityManager()->MarkForRemoval(netEntityHandle);*/
+
+                // Until https://github.com/o3de/o3de/issues/11695 is fixed, move the gem out of the view
```

Comment:
This does make me wonder if we should consider pooling

Reply:
We definitely can but this is also a valid approach that we should test. But I agree, for another system in the sample, we should show to do pooling for network objects.